pkg/util/pki/csr.go

```diff
@@ -164,11 +163,33 @@ func BuildCertManagerKeyUsages(ku x509.KeyUsage, eku []x509.ExtKeyUsage) []v1.Ke
 	return usages
 }
 
+type generateCSROptions struct {
+	EncodeBasicConstraintsInRequest bool
+}
+
+type GenerateCSROption func(*generateCSROptions)
+
+// WithEncodeBasicConstraintsInRequest determines whether the BasicConstraints
+// extension should be encoded in the CSR.
+// NOTE: this is a temporary option that will be removed in a future release.
+func WithEncodeBasicConstraintsInRequest(encode bool) GenerateCSROption {
+	return func(o *generateCSROptions) {
+		o.EncodeBasicConstraintsInRequest = encode
+	}
+}
+
 // GenerateCSR will generate a new *x509.CertificateRequest template to be used
 // by issuers that utilise CSRs to obtain Certificates.
 // The CSR will not be signed, and should be passed to either EncodeCSR or
 // to the x509.CreateCertificateRequest function.
-func GenerateCSR(crt *v1.Certificate) (*x509.CertificateRequest, error) {
+func GenerateCSR(crt *v1.Certificate, optFuncs ...GenerateCSROption) (*x509.CertificateRequest, error) {
+	opts := &generateCSROptions{
+		EncodeBasicConstraintsInRequest: false,
+	}
+	for _, opt := range optFuncs {
+		opt(opts)
+	}
```

Comment:
question: Are we actually going to remove `WithEncodeBasicConstraintsInRequest` in the future? I'd say more likely is that we default to it being `true` rather than removing it. Then if it causes someone a problem they have remediation available (by calling `WithEncodeBasicConstraintsInRequest(false)`)

suggestion: If we _are_ going to change the default, would it be worth mentioning that it could change in the doc comment for `GenerateCSR`?

Reply:
I think that in the future we should add a boolean option in the Certificate resource and this option should be replaced with a `DefaultEncodeBasicConstraintsInRequest` option which controls what the default behavior is in case the boolean field is not set.